modules/demo/client-server/routes/cudf.js

```diff
@@ -0,0 +1,277 @@
+}
+
+/**
+ * Computes cudf.DataFrame.filter by parsing input query_dictionary, and performing filterByValue or
+ * filterByRange as needed.
+ *
+ * @param query_dict {}, where key is column name, and value is either scalar value, or a range of
+ *   scalar values
+ * @param ignore str, column name which is to be ignored (if any) from query compute
+ * @returns resulting cudf.DataFrame
+ */
+function parseQuery(query_dict, ignore) {
+  if (ignore in query_dict) { delete query_dict[ignore]; }
+  var t0       = new Date().getTime();
+  let boolmask = undefined;
+  for (const [key, value] of Object.entries(query_dict)) {
+    let temp_boolmask = undefined;
+    if (df._accessor.names.includes(key)) {
+      if (Array.isArray(value) && value.length == 2) {
+        temp_boolmask = filterByRange(key, value[0], value[1]);
+      } else if (typeof (value) == 'number') {
+        temp_boolmask = filterByValue(key, value);
+      }
+    }
+    if (temp_boolmask !== undefined) {
+      boolmask = (boolmask == undefined) ? temp_boolmask : boolmask.logical_and(temp_boolmask);
+    }
+  }
+  var res = '';
+  if (boolmask == undefined) {
+    res = df;
+  } else {
+    res = df.filter(boolmask);
+  }
+  var t1 = new Date().getTime();
+  console.log('Query ', query_dict, ', Time Taken:', t1 - t0, 'ms');
+  return res;
+}
+
+/**
+ * Computes cudf.DataFrame.groupby.
+ *
+ * @param df cudf.DataFrame object
+ * @param by str, column name by which the groupby needs to be computed
+ * @param agg str, aggregate function, one of the following:
+ *   https://rapidsai.github.io/node-rapids/classes/cudf_src.groupbysingle.html#methods
+ * @returns resulting cudf.DataFrame
+ */
+function groupby(df, by, agg) {
+  var t0         = new Date().getTime();
+  const grp      = df.groupBy({by: by});
+  const validAgg = (agg in grp && typeof grp[agg] == 'function');
+  if (!validAgg) { return ({}); }
+  var res = grp[agg]();
+  var t1  = new Date().getTime();
+  console.log('Group by ', by, ', agg:', agg, 'Time Taken:', t1 - t0, 'ms');
+  return res;
+}
+
+// socket connection
+module.exports = (io) => {
+  router.get('/', (req, res) => { res.send('ok'); })
+
+  // SOCKET.IO paths
+  io.on('connection', (socket) => {
+    console.log('client connected');
+
+    socket.on('disconnect', () => { console.log('client disconnected'); });
+
+    socket.on('readMortgageData', (callback) => {
+      if (datasetLoaded == 'mortgage') {
+        console.log('dataset already loaded');
+        socket.emit('data-points-update', df.toArrow().length);
+        callback(true);
+      } else {
+        readMortgageData((cb) => {
+          if (cb == true) {
+            socket.emit('data-points-update', df.toArrow().length);
+            callback(true);
+          }
+        });
+      }
+    });
+
+    socket.on('readUberData', (callback) => {
+      if (datasetLoaded == 'uber') {
+        console.log('dataset already loaded');
+        socket.emit('data-points-update', df.toArrow().length);
```

Comment:
```suggestion
        socket.emit('data-points-update', df.length);
```

Reply:
@trxcllnt `df.length` is undefined! 

`df` is a cudf DataFrame here